### PR TITLE
Refactor conv

### PR DESCRIPTION
**Pull request by rongzha1**
## Description ##
modify class MKLDNNConvForward to new structure

### Essentials ###
- [done] Passed code style checking (`make lint`)
- [done] Changes are complete (i.e. I finished coding on this PR)
- [done] All changes have test coverage
- [done] For user-facing API changes, API doc string has been updated. For new C++ functions in header files, their functionalities and arguments are well-documented. 
- [done] To my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [done] Feature1, tests, (and when applicable, API doc)
modify class MKLDNNConvForward to new structure

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 1]
```diff
@@ -313,9 +312,13 @@ class MKLDNNOpSignature {
...
-    hash = hash * 2 + arr.dtype();
-    eles.push_back(arr.dtype());
-    AddSign(arr.shape());
+    if (arr.IsMKLDNN()) {
+      AddSign(*(arr.GetMKLDNNData()));
+    } else {
+      hash = hash * 2 + arr.dtype();
+      eles.push_back(arr.dtype());
+      AddSign(arr.shape());
+    }
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: this code has been merged. please remove this.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 2]
```diff
@@ -380,14 +383,16 @@ typedef std::pair<OutDataOp, mkldnn::memory *> mkldnn_output_t;
  * The difference is that the first function can create MKLDNN memory with
  * special layouts in an NDArray, while the second one can only create MKLDNN
  * memory with default layouts.
+ * If these two functions are used, we have to call CommitOutput to write
+ * the output back to the output NDArray.
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: these comments have been added to my branch. please remove them.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 3]
```diff
@@ -380,14 +383,16 @@ typedef std::pair<OutDataOp, mkldnn::memory *> mkldnn_output_t;
...
  */
 mkldnn_output_t CreateMKLDNNMem(const NDArray &arr,
                                 const mkldnn::memory::primitive_desc &desc,
                                 OpReqType req);
 mkldnn_output_t CreateMKLDNNWeightGrad(const NDArray &arr,
                                        const mkldnn::memory::primitive_desc &desc,
                                        OpReqType req);
-
+/* This function has to be used with one of the functions above. */
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: this one as well.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 4]
```diff
@@ -398,7 +403,12 @@ mkldnn_memory_format_t GetDefaultFormat(mkldnn::memory::desc desc);
 mkldnn::memory::primitive_desc GetPrimitiveDesc(mkldnn::memory::primitive_desc pd,
                                                 mkldnn_memory_format_t format);
 
+mkldnn::convolution_forward::primitive_desc GetConvFwdPd(int dilate_dim, const NDArray *bias,
+    bool is_train, const mkldnn::memory::desc &data_md, const mkldnn::memory::desc &weight_md,
+    const mkldnn::memory::desc &out_md, const mkldnn::memory::dims &strides,
+    const mkldnn::memory::dims &dilates, const mkldnn::memory::dims &padding);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: do we need to declare it here?

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 5]
```diff
@@ -151,9 +151,12 @@ const mkldnn::memory *GetWeights(const NDArray &arr,
...
     return nullptr;
   }
-  if (mem == nullptr)
+  if (mem == nullptr) {
     mem = arr.GetMKLDNNDataReorder(target_pd);
-  if (mem->get_primitive_desc() == target_pd) return mem;
+  }
+  if (mem->get_primitive_desc() == target_pd) {
+    return mem;
+  }
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: please avoid changing the way the original code was written if you are not modifying it.

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 6]
```diff
@@ -234,6 +237,42 @@ mkldnn::memory::primitive_desc GetPrimitiveDesc(mkldnn::memory::primitive_desc p
...
+      auto bias_md = GetMemDesc(*bias);
+      mkldnn::convolution_forward::desc desc(prop, mkldnn::algorithm::convolution_direct,
+                                             data_md, weight_md, bias_md, out_md, strides,
+                                             dilates, padding, padding,
+                                             mkldnn::padding_kind::zero);
+      return mkldnn::convolution_forward::primitive_desc(desc, engine);
+    }
+  }
+}
+
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: why define it here?

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 7]
```diff
@@ -38,49 +38,37 @@ static mkldnn::convolution_forward::primitive_desc GetConvFwdImpl(
...
+    LOG(FATAL) << "Unsupported pad dim";
+  }
+
+  mkldnn::memory::dims dilates{0, 0};
+  if (param.dilate.ndim() == 2) {
+    dilates[0] = param.dilate[0] - 1;
+    dilates[1] = param.dilate[1] - 1;
   }
+  return GetConvFwdPd(param.dilate.ndim(), bias, is_train, data_md, weight_md, out_md
+    , strides, dilates, padding);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: please follow the code style.

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 8]
```diff
@@ -127,22 +125,35 @@ static mkldnn::convolution_backward_weights::primitive_desc GetConvBwdWeights(
...
+      mkldnn::convolution_backward_weights::desc desc(mkldnn::algorithm::convolution_direct,
+          data_md, weight_md, out_md, strides, padding, padding, mkldnn::padding_kind::zero);
+      return mkldnn::convolution_backward_weights::primitive_desc(desc, engine, fwd_pd);
+    } else {
+      auto bias_md = GetMemDesc(*bias);
+      mkldnn::convolution_backward_weights::desc desc(mkldnn::algorithm::convolution_direct,
+          data_md, weight_md, bias_md, out_md, strides, padding, padding,
+          mkldnn::padding_kind::zero);
+      return mkldnn::convolution_backward_weights::primitive_desc(desc, engine, fwd_pd);
+    }
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: why do you rewrite the code here? are you adding something here? if not, please change the code back. it's difficult for me to review your code.

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 9]
```diff
@@ -177,54 +188,88 @@ class MKLDNNConvForward {
...
+  void SetDataHandle(const nnvm::NodeAttrs &attrs,
+      const OpContext &ctx, const std::vector<NDArray> &in_data,
+      const std::vector<OpReqType> &req, const std::vector<NDArray> &out_data);
+
+  void Execute(const std::vector<OpReqType> &req, const std::vector<NDArray> &out_data);
+};  // class MKLDNNConvForward
+
+MKLDNNConvForward::MKLDNNConvForward(const ConvolutionParam& param, bool is_train,
+                      const NDArray &data, const NDArray &weights,
+                      const NDArray *bias, const NDArray &output): fwd_pd(
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: fix indent here.

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 10]
```diff
@@ -177,54 +188,88 @@ class MKLDNNConvForward {
...
+    this->fwd = std::shared_ptr<mkldnn::convolution_forward>(
+      new mkldnn::convolution_forward(fwd_pd, mkldnn::primitive::at(*this->data),
+                                      mkldnn::primitive::at(*this->weight),
+                                      *this->out));
   }
+}
+
+void MKLDNNConvForward::SetDataHandle(const nnvm::NodeAttrs &attrs,
+    const OpContext &ctx, const std::vector<NDArray> &in_data,
+    const std::vector<OpReqType> &req, const std::vector<NDArray> &out_data) {
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: fix indent here.

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 11]
```diff
@@ -177,54 +188,88 @@ class MKLDNNConvForward {
...
+    if (!param.no_bias) {
+      bias_mem = in_data[conv::kBias].GetMKLDNNDataReorder(this->fwd_pd.bias_primitive_desc());
+        this->bias->set_data_handle(bias_mem->get_data_handle());
+      }
+}
+
+void MKLDNNConvForward::Execute(const std::vector<OpReqType> &req,
+                                   const std::vector<NDArray> &out_data) {
+  auto out_mem = CreateMKLDNNMem(out_data[conv::kOut], this->fwd_pd.dst_primitive_desc(),
+                                 req[conv::kOut]);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: don't call CreateMKLDNNMem twice.
